File: src/main/java/com/melonbase/microquark/rest/mapping/LocalDateParameterConverterProvider.java
package com.melonbase.microquark.rest.mapping;

import javax.ws.rs.ext.ParamConverter;
import javax.ws.rs.ext.ParamConverterProvider;
import javax.ws.rs.ext.Provider;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.time.LocalDate;

@Provider
public class LocalDateParameterConverterProvider implements ParamConverterProvider {

  @Override
  public <T> ParamConverter<T> getConverter(
      final Class<T> rawType,
      final Type genericType,
      final Annotation[] annotations
  ) {
    if (LocalDate.class.equals(rawType)) {
      return (ParamConverter<T>) new LocalDateParamConverter();
    }
    return null;
  }
}

File: src/main/kotlin/com/melonbase/microquark/rest/mapping/LocalDateParameterConverterProvider.kt
package com.melonbase.microquark.rest.mapping

import java.lang.reflect.Type
import java.time.LocalDate
import javax.ws.rs.ext.ParamConverter
import javax.ws.rs.ext.ParamConverterProvider
import javax.ws.rs.ext.Provider

/*
 * Obsolete when Quarkus 1.13 is released
 * https://github.com/quarkusio/quarkus/pull/15451
 */
@Provider
class LocalDateParameterConverterProvider : ParamConverterProvider {

  override fun <T> getConverter(
    rawType: Class<T>,
    genericType: Type,
    annotations: Array<Annotation>
  ): ParamConverter<T>? {
    return if (LocalDate::class.java == rawType) {
      @Suppress("UNCHECKED_CAST")
      LocalDateParamConverter() as ParamConverter<T>
    } else null
  }
}